src/Callable.h

```diff
@@ -0,0 +1,357 @@
+    }
+
+    static bool is_compatible_fcci(FullCallCheckInfo actual, FullCallCheckInfo expected) {
+        if (actual == expected) {
+            return true;  // my, that was easy
+        }
+
+        // Might still be compatible
+        const bool a_is_buffer = (actual.lanes & 1) != 0;
+        const int a_dims = (((int16_t)actual.lanes) >> 1);
+        const halide_type_t a_type = actual.with_lanes(0);
+
+        const bool e_is_buffer = (expected.lanes & 1) != 0;
+        const int e_dims = (((int16_t)expected.lanes) >> 1);
+        const halide_type_t e_type = expected.with_lanes(0);
+
+        const bool types_match = (a_type == halide_type_t()) ||
+                                 (e_type == halide_type_t()) ||
+                                 (a_type == e_type);
+
+        const bool dims_match = a_dims < 0 ||
+                                e_dims < 0 ||
+                                a_dims == e_dims;
+
+        return a_is_buffer == e_is_buffer && types_match && dims_match;
+    }
+
+    template<typename T>
+    static constexpr FullCallCheckInfo make_fcci() {
+        using T0 = typename std::remove_const<typename std::remove_reference<T>::type>::type;
+        if constexpr (Internal::IsHalideBuffer<T0>::value) {
+            using TypeAndDims = Internal::HalideBufferStaticTypeAndDims<T0>;
+            return make_buffer_fcci(TypeAndDims::type(), TypeAndDims::dims());
+        } else if constexpr (std::is_arithmetic<T0>::value || std::is_pointer<T0>::value) {
+            return make_scalar_fcci(halide_type_of<T0>());
+        } else {
+            // static_assert(false) will fail all the time, even inside constexpr,
+            // but gating on sizeof(T) is a nice trick that ensures we will always
+            // fail here (since no T is ever size 0).
+            static_assert(!sizeof(T), "Illegal type passed to Callable.");
+        }
+    }
+
+    template<typename... Args>
+    static constexpr std::array<FullCallCheckInfo, sizeof...(Args)> make_fcci_array() {
+        return std::array<FullCallCheckInfo, sizeof...(Args)>{make_fcci<Args>()...};
+    }
+
+    // ---------------------------------
+
+    template<int Size>
+    struct ArgvStorage {
+        const void *argv[Size];
+        // We need a place to store the scalar inputs, since we need a pointer
+        // to them and it's better to avoid relying on stack spill of arguments.
+        // Note that this will usually have unused slots, but it's cheap and easy
+        // compile-time allocation on the stack.
+        uintptr_t argv_scalar_store[Size];
+
+        template<typename... Args>
+        explicit ArgvStorage(Args &&...args) {
+            fill_slots(0, std::forward<Args>(args)...);
+        }
+
+    private:
+        template<typename T, int Dims>
+        HALIDE_ALWAYS_INLINE void fill_slot(size_t idx, const ::Halide::Buffer<T, Dims> &value) {
+            // Don't call ::Halide::Buffer::raw_buffer(): it includes "user_assert(defined())"
```

Comment:
So what happens if you pass an undefined buffer?

I don't think we should worry too much about the cost of easily-predicted branches. They're often free.

Reply:
You crash on a null pointer access. Don't do that :-)

But seriously, I'll include the user_assert() if you like, but it included a surprising amount of code (including an out-of-line call IIRC), enough to make me want to elide it here -- I was really trying hard to keep the amount of hidden overhead to a bare minimum. Let me know if I should re-add it.

Reply:
If we could somehow convert that case into passing a nullptr halide_buffer_t as the argument, then the result would be the Halide pipeline throwing an error. E.g:
```
auto ptr = value.get();
ptr ? ptr->raw_buffer : nullptr;
```

Reply:
Nice compromise, done.